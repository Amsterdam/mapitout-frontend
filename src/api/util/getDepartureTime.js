export const getDepartureTime = (date = new Date()) => {
  const dayOfWeek = date.getUTCDay();

  if (dayOfWeek !== 1) {
    const dayOfMonth = date.getUTCDate();

    date.setUTCDate(dayOfMonth - dayOfWeek + 8);
  }

  date.setUTCHours(9, 0, 0, 0);

  return date;
};
