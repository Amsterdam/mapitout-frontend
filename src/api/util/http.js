export const request = async (url, requestConfig) => {
  let response;

  response = await fetch(url, requestConfig);

  if (response.ok) {
    try {
      return await response.json();
    } catch (error) {
      throw new Error("Invalid server response");
    }
  } else {
    throw new Error("Invalid server response");
  }
};
