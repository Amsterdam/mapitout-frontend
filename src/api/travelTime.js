import { request } from "./util/http";
import { omit } from "lodash-es";
import { ENDPOINT_RANGES } from "../env";
import { getDepartureTime } from "./util/getDepartureTime";

export const cache = {
  areas: {}
};

export const fetchAreas = async (rangesParams = []) => {
  if (rangesParams instanceof Array === false || rangesParams.length === 0) {
    return [];
  }
  const cacheKey = rangesParams
    .map(rangeParams => omit(rangeParams, ["id", "origin"]))
    .map(rangeParams =>
      Object.values(rangeParams)
        .sort()
        .join("-")
    )
    .join(";");

  if (typeof cache.areas[cacheKey] !== "undefined") {
    return cache.areas[cacheKey];
  }

  const requestBody = {
    departure_searches: rangesParams.map(rangeParams => {
      return {
        id: rangeParams.id,
        coords: {
          lat: rangeParams.originLat,
          lng: rangeParams.originLng
        },
        departure_time: getDepartureTime(),
        travel_time: rangeParams.travelTime * 60,
        transportation: {
          type: rangeParams.transportTypeValue
        }
      };
    }),
    unions: [
      {
        id: "union",
        search_ids: rangesParams.map(rangeParams => rangeParams.id)
      }
    ]
  };

  const url = new URL(ENDPOINT_RANGES);

  const response = await request(url.toString(), {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-type": "application/json; charset=utf-8"
    },
    body: JSON.stringify(requestBody)
  });

  const result = response.results.map(timeMap => {
    return {
      id: timeMap.search_id,
      paths: timeMap.shapes.reduce((acc, shape) => {
        let paths = [shape.shell];

        if (shape.holes.length > 0) {
          paths = paths.concat(shape.holes);
        }

        return acc.concat(paths);
      }, [])
    };
  });

  cache.areas[cacheKey] = result;

  return result;
};
