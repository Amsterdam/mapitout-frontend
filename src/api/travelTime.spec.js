import { fetchAreas, cache } from "./travelTime";
import * as http from "./util/http";
import { omit } from "lodash-es";

import * as getDepartureTimeModule from "./util/getDepartureTime";

afterEach(() => {
  jest.resetAllMocks();
});

jest.mock("../env.js");

const requestSpy = jest.spyOn(http, "request").mockImplementation();

describe("fetchAreas", () => {
  beforeEach(() => {
    cache.areas = {};
  });

  const getDepartureTimeSpy = jest
    .spyOn(getDepartureTimeModule, "getDepartureTime")
    .mockImplementation();

  beforeEach(() => {
    getDepartureTimeSpy.mockReturnValue(new Date("2019-07-15T09:00:00.000Z"));
  });

  const validRangeParam1 = {
    id: "0",
    origin: "",
    originId: "",
    originLat: 0,
    originLng: 0,
    originTypeId: 0,
    travelTime: 45,
    transportTypeValue: "0"
  };

  const validRangeParam2 = {
    id: "1",
    origin: "",
    originId: "",
    originLat: 0,
    originLng: 0,
    originTypeId: 0,
    travelTime: 45,
    transportTypeValue: "0"
  };

  it("should return an empty array when passed an invalid or empty rangesParams parameter", async () => {
    const result1 = await fetchAreas({});
    const result2 = await fetchAreas([]);

    expect(result1).toEqual([]);
    expect(result2).toEqual([]);
    expect(requestSpy).toHaveBeenCalledTimes(0);
  });

  it("should return a cached value if any is found", async () => {
    const cacheKey = [validRangeParam1]
      .map(rangeParams => omit(rangeParams, ["id", "origin"]))
      .map(rangeParams =>
        Object.values(rangeParams)
          .sort()
          .join("-")
      )
      .join(";");

    const expectedResult = { results: [] };

    cache.areas[cacheKey] = expectedResult;

    const result = await fetchAreas([validRangeParam1]);

    expect(requestSpy).toHaveBeenCalledTimes(0);
    expect(result).toEqual(expectedResult);
  });

  describe("when no cached value exists", () => {
    it("should call request with the correct url and request configuration object", async () => {
      const rangesParams1 = [validRangeParam1];
      const rangesParams2 = [validRangeParam1, validRangeParam2];

      requestSpy.mockResolvedValue({
        results: []
      });

      await fetchAreas(rangesParams1);
      await fetchAreas(rangesParams2);

      getDepartureTimeSpy.mockReturnValue(new Date("2019-07-15T09:00:00.000Z"));

      expect(requestSpy).toHaveBeenCalledTimes(2);
      expect(requestSpy.mock.calls[0]).toMatchSnapshot();
      expect(requestSpy.mock.calls[1]).toMatchSnapshot();
    });

    it("should throw an error when request rejects", async () => {
      const expectedError = new Error("test-error");

      requestSpy.mockRejectedValue(expectedError);

      await expect(fetchAreas([validRangeParam1])).rejects.toThrow(
        expectedError
      );
    });

    it("should throw and error if request returns an invalid response", async () => {
      requestSpy.mockResolvedValue({
        response: {}
      });

      await expect(fetchAreas([validRangeParam1])).rejects.toThrow();
    });

    it("should return an empty array if request returns an empty response", async () => {
      const expectedResult = [];

      requestSpy.mockResolvedValue({
        results: []
      });

      const result = await fetchAreas("test");

      expect(result).toEqual(expectedResult);
    });

    it("should return the correct value when request resolves with a valid response and save it to cache", async () => {
      const rangesParams = [validRangeParam1, validRangeParam2];

      requestSpy.mockResolvedValue({
        results: [
          {
            search_id: "0",
            rangeId: "0",
            shapes: [
              {
                shell: [[1, 2]],
                holes: [[3, 4]]
              }
            ]
          }
        ]
      });

      const result = await fetchAreas(rangesParams);

      const cacheKey = rangesParams
        .map(rangeParams => omit(rangeParams, ["id", "origin"]))
        .map(rangeParams =>
          Object.values(rangeParams)
            .sort()
            .join("-")
        )
        .join(";");

      const expectedResult = [{ id: "0", paths: [[[1, 2]], [3, 4]] }];

      expect(result).toEqual(expectedResult);
      expect(cache.areas[cacheKey]).toEqual(expectedResult);
    });
  });
});
