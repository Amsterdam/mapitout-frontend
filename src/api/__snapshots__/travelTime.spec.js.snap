// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAreas when no cached value exists should call request with the correct url and request configuration object 1`] = `
Array [
  "https://www.example.com/ranges",
  Object {
    "body": "{\\"departure_searches\\":[{\\"id\\":\\"0\\",\\"coords\\":{\\"lat\\":0,\\"lng\\":0},\\"departure_time\\":\\"2019-07-15T09:00:00.000Z\\",\\"travel_time\\":2700,\\"transportation\\":{\\"type\\":\\"0\\"}}],\\"unions\\":[{\\"id\\":\\"union\\",\\"search_ids\\":[\\"0\\"]}]}",
    "headers": Object {
      "Accept": "application/json",
      "Content-type": "application/json; charset=utf-8",
    },
    "method": "POST",
  },
]
`;

exports[`fetchAreas when no cached value exists should call request with the correct url and request configuration object 2`] = `
Array [
  "https://www.example.com/ranges",
  Object {
    "body": "{\\"departure_searches\\":[{\\"id\\":\\"0\\",\\"coords\\":{\\"lat\\":0,\\"lng\\":0},\\"departure_time\\":\\"2019-07-15T09:00:00.000Z\\",\\"travel_time\\":2700,\\"transportation\\":{\\"type\\":\\"0\\"}},{\\"id\\":\\"1\\",\\"coords\\":{\\"lat\\":0,\\"lng\\":0},\\"departure_time\\":\\"2019-07-15T09:00:00.000Z\\",\\"travel_time\\":2700,\\"transportation\\":{\\"type\\":\\"0\\"}}],\\"unions\\":[{\\"id\\":\\"union\\",\\"search_ids\\":[\\"0\\",\\"1\\"]}]}",
    "headers": Object {
      "Accept": "application/json",
      "Content-type": "application/json; charset=utf-8",
    },
    "method": "POST",
  },
]
`;
